Add documentation for min_hash filter

mayya-sharipova opened the pull request:
Closes #20757

mayya-sharipova committed:
Add documentation for min_hash filter

Closes #20757

elasticmachine commented:
Pinging @elastic/es-search

cbuescher reviewed (commented):
Great addition to the docs, I left some very minor comments. 
One general thought I was having: I understand why it makes sense to start with a sort of "overview" and theory, but since or docs also work as a kind of reference guide, maybe we should aim for a very brief summary (like the existing one, maybe extended slightly) followed by the table of parameters, then add the more detailed "theory" and usage sections afterwards.
Also I was wondering if it would make sense to add a small example of how to actually use any such min-hashed field in a query, e.g. for near duplicate detecton etc... or if this would go beyond the scope of our documentation.

cbuescher commented on docs/reference/analysis/tokenfilters/minhash-tokenfilter.asciidoc:
```diff
@@ -20,3 +54,80 @@ The following are settings that can be set for a `min_hash` token filter.
+  }
+}
+--------------------------------------------------
+// NOTCONSOLE
```
Maybe add some short callouts linking the sections and parameters to what has been explained in the test before. Also I would probably either rename the field to something more specific than "text" (e.g. hashed_text, fingerprint, minhash etc...) to make clear the contents of this field is not what you would typically search on.

mayya-sharipova commented on docs/reference/analysis/tokenfilters/minhash-tokenfilter.asciidoc:
```diff
@@ -20,3 +54,80 @@ The following are settings that can be set for a `min_hash` token filter.
+internally each shingle is hashed into to 128-bit hash, you should choose
+`k` small enough so that all possible
+different k-words shingles can be hashed to 128-bit hash with
+minimal collision. 5-word shingles typically work well.
```
@cbuescher  I took an advice on `5-word shingle` from the MinHash filter [sourcecode](https://github.com/apache/lucene-solr/blob/master/lucene/analysis/common/src/java/org/apache/lucene/analysis/minhash/MinHashFilter.java#L40) in Lucene

cbuescher replied:
Thats interesting, would you mind linking to that source?

mayya-sharipova replied:
@cbuescher Thanks for the suggetion. I opted not to include  the link to this source, as I am afraid as the sourcecode changes this link becomes invalid.

jtibshirani replied:
In the original PR that adds `min_hash`, it looks like we were not sure about the 5 word suggestion, and instead encouraged 2 word shingles: https://github.com/elastic/elasticsearch/pull/20206#issuecomment-243129857. It would be nice if there was a reference or set of experiments to help confirm a good default value... I didn't manage to find one in a quick search, but will keep a lookout. The right choice seems like it would depend on the use case as well (for example similarity search vs. duplicate detection).

mayya-sharipova replied:
@jtibshirani  Thanks a lot for the review.  I think the best for now is to remove this line completely "5-word shingles typically work well.", as there are conflicting suggestions what shingle size works best.  Once we have better sources (external or from our own experiments), we can add shingle size suggestions to the file. Is this fine with you?

jtibshirani replied:
This sounds like a good plan to me!

mayya-sharipova committed:
Address feedback

mayya-sharipova commented:
@cbuescher Thanks a lot for the review. I have addressed your comments in the 2nd commit.

> Also I was wondering if it would make sense to add a small example of how to actually use any such min-hashed field in a query, e.g. for near duplicate detecton etc... or if this would go beyond the scope of our documentation.

I indeed very much wanted to add this example, but I opted not to do this. The reason for this that I am not sure how to set the best query for this. A [general idea ](https://www.youtube.com/watch?v=_1D35bN95Go) is to partition resulting hashed tokens into bands; tokens in a single band should be joined by `AND`, and bands should be joined with other bands by `OR`.  I have [asked the author](https://issues.apache.org/jira/browse/LUCENE-6968?focusedCommentId=16783921&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16783921) of the MinHash filter for his idea how this query should be set. When he replies, we can update the documentation with this query information as well.

cbuescher reviewed (approved):
@mayya-sharipova  Thanks for adressing my previous comments, I left one more suggestion but nothing that requires another review. Feel free to adress or not.

> When he replies, we can update the documentation with this query information as well

Fine by me, this PR already is a great addition. Maybe an extended example would also be better suited for a blog post or something like it. I'd be really interested in real-life usages of this.

jtibshirani reviewed (commented):
I just read over the documentation to learn more about this token filter and had a couple thoughts. I found these additions very helpful!

rdvdijk commented:
The documentation does not show or explain how to query for `min_hash` values.

For others that are trying to figure out how to query for `min_hash` values, and end up here (as I did), an explanation on how to do this can be found in the comments of the issue: https://github.com/elastic/elasticsearch/issues/20757#issuecomment-1196488373

I think it would be a good idea to add such an example to the documentation.
